### PR TITLE
[IMP] project_timesheet_holidays: prevent modifying leave timesheets

### PR DESCRIPTION
**Pull request by alexey-pelykh**
Description of the issue/feature this PR addresses:
#30850

Current behavior before PR:
 * Can modify timesheet lines on approved leaves

Desired behavior after PR is merged:
 * Can not modify timesheet lines on approved leaves

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by pedrobaeza** [comment]
You have making several changes in the same PR. Better to address little things each time. You are adding a hook and also avoiding data modification. Do only one here and the other in another PR (although hooks are not usually accepted in stable).

**Comment by alexey-pelykh** [comment]
@pedrobaeza no hooks here - not sure I'm on the same page, this PR only refactors code and adds `write()` as well as addresses privacy issue + it's for `master`, such PR won't make it to `12.0` I guess

File: addons/project_timesheet_holidays/models/hr_holidays.py [hunk 1]
```diff
@@ -55,6 +55,36 @@ class Holidays(models.Model):
...
+            self.date_to,
+        )
+
+        for index, (day_date, work_hours_count) in enumerate(work_hours_data):
+            self.env['account.analytic.line'].create(
+                self._create_timesheet_line_data(index, work_hours_data, day_date, work_hours_count)
+            )
+
+    @api.multi
+    def _create_timesheet_line_data(self, index, work_hours_data, day_date, work_hours_count):
```
Comment (pedrobaeza) [inline addons/project_timesheet_holidays/models/hr_holidays.py]: This is the hook I'm mentioning.

Comment (alexey-pelykh) [reply addons/project_timesheet_holidays/models/hr_holidays.py]: Ah, got it. Well, I can move it back, though other modules in Odoo use same approach.

Comment (alexey-pelykh) [reply addons/project_timesheet_holidays/models/hr_holidays.py]: Let me rephrase - if this going to be the reason of rejection - I'll move it back, since it was added w/o any particular usage in mind at least right now

Comment (pedrobaeza) [reply addons/project_timesheet_holidays/models/hr_holidays.py]: Yes, as said, 1 change each time.

Comment (alexey-pelykh) [reply addons/project_timesheet_holidays/models/hr_holidays.py]: Done, now there's #30867

File: addons/project_timesheet_holidays/models/account_analytic.py [hunk 2]
```diff
@@ -10,8 +10,27 @@ class AccountAnalyticLine(models.Model):
...
+        return super().write(values)
+
     @api.multi
     def unlink(self):
-        if any(line.holiday_id for line in self):
-            raise UserError(_('You cannot delete timesheet lines attached to a leaves. Please cancel the leaves instead.'))
-        return super(AccountAnalyticLine, self).unlink()
+        if self.filtered('holiday_id'):
+            raise UserError(_('You cannot delete timesheet lines attached to a leave. Please cancel the leave instead.'))
+        return super().unlink()
```
Comment (mreficent) [inline addons/project_timesheet_holidays/models/account_analytic.py]: super... 😅

File: addons/project_timesheet_holidays/tests/test_timesheet_holidays.py [hunk 3]
```diff
@@ -104,3 +105,19 @@ def test_validate_without_timesheet(self):
...
+            'employee_id': self.empl_employee.id,
+            'holiday_status_id': self.hr_leave_type_with_ts.id,
+            'date_from': self.leave_start_datetime,
+            'date_to': self.leave_end_datetime,
+            'number_of_days': number_of_days,
+        })
+        holiday.sudo().action_validate()
+
+        with self.assertRaises(UserError):
+            holiday.timesheet_ids.unit_amount = 4
```
Comment (mreficent) [inline addons/project_timesheet_holidays/tests/test_timesheet_holidays.py]: runbot fails here, fix:
```python
holiday.timesheet_ids[0].unit_amount = 4
```

Comment (alexey-pelykh) [reply addons/project_timesheet_holidays/tests/test_timesheet_holidays.py]: Thanks, adjusted

File: addons/project_timesheet_holidays/tests/test_timesheet_holidays.py [hunk 4]
```diff
@@ -104,3 +105,19 @@ def test_validate_without_timesheet(self):
         })
         holiday.sudo().action_validate()
         self.assertEquals(len(holiday.timesheet_ids), 0, 'Number of generated timesheets should be zero since the leave type does not generate timesheet')
+
+    def test_error_on_modification(self):
```
Comment (jem-odoo) [inline addons/project_timesheet_holidays/tests/test_timesheet_holidays.py]: dosstring explaining the test

**Review by jem-odoo — CHANGES_REQUESTED** [review]
Coding review. Interesting feature for master.

Please detail a little the use case motivation and consequence in your commit message. The commit title should not be "ensure data integrity" but "prevent modifying leave timesheets"

File: addons/project_timesheet_holidays/tests/test_timesheet_holidays.py (hunk 4 above)
Comment (alexey-pelykh) [reply addons/project_timesheet_holidays/tests/test_timesheet_holidays.py]: done

File: addons/project_timesheet_holidays/models/account_analytic.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (alexey-pelykh) [reply addons/project_timesheet_holidays/models/account_analytic.py]: done

Comment (alexey-pelykh) [reply addons/project_timesheet_holidays/models/account_analytic.py]: @mreficent which one?

File: addons/project_timesheet_holidays/models/account_analytic.py [hunk 5]
```diff
@@ -10,8 +10,27 @@ class AccountAnalyticLine(models.Model):
...
+                    'account_id',
+                    'unit_amount',
+                    'user_id',
+                    'date',
+                    'employee_id',
+                ]):
+            raise UserError(_(
+                'You cannot modify timesheet lines attached to a leave.'
+            ))
+        return super().write(values)
```
[the remark this answers is not in the document]
Comment (alexey-pelykh) [reply addons/project_timesheet_holidays/models/account_analytic.py]: done

File: addons/project_timesheet_holidays/models/account_analytic.py [hunk 6]
```diff
@@ -10,8 +10,27 @@ class AccountAnalyticLine(models.Model):
...
+                    'project_id',
+                    'task_id',
+                    'account_id',
+                    'unit_amount',
+                    'user_id',
+                    'date',
+                    'employee_id',
+                ]):
+            raise UserError(_(
+                'You cannot modify timesheet lines attached to a leave.'
```
[the remark this answers is not in the document]
Comment (alexey-pelykh) [reply addons/project_timesheet_holidays/models/account_analytic.py]: done

File: addons/project_timesheet_holidays/models/account_analytic.py [hunk 7]
```diff
@@ -10,8 +10,27 @@ class AccountAnalyticLine(models.Model):
 
     holiday_id = fields.Many2one("hr.leave", string='Leave Request')
 
+    @api.multi
+    def write(self, values):
+        if not self.env.context.get('skip_leave_integrity_check', False) and \
+                self.filtered('holiday_id') and \
+                any(field in values for field in [
```
[the remark this answers is not in the document]
Comment (alexey-pelykh) [reply addons/project_timesheet_holidays/models/account_analytic.py]: done

File: addons/project_timesheet_holidays/models/account_analytic.py [hunk 8]
```diff
@@ -10,8 +10,27 @@ class AccountAnalyticLine(models.Model):
 
     holiday_id = fields.Many2one("hr.leave", string='Leave Request')
 
+    @api.multi
+    def write(self, values):
+        if not self.env.context.get('skip_leave_integrity_check', False) and \
```
[the remark this answers is not in the document]
Comment (alexey-pelykh) [reply addons/project_timesheet_holidays/models/account_analytic.py]: done

**Comment by jem-odoo** [comment]
Note for myself: prevent leave timesheet modification can cause problem when adjust a timesheet in grid view. We will need to exlude those kind of timesheet, which implies a bridge module between timesheet_grid and project_timesheet_holidays.

**Comment by alexey-pelykh** [comment]
@jem-odoo do I need to fix the conflicts?

**Comment by jem-odoo** [comment]
@alexey-pelykh I will have to discuss with the Product Owner if it is worthy to handle that use case.

**Comment by alexey-pelykh** [comment]
@jem-odoo any feedback available?

**Comment by alexey-pelykh** [comment]
It's impressive how a PR can be unattended for 6 years without any feedback
